src/main/flight/pid.c

```diff
@@ -292,13 +315,99 @@ static void pidApplyRateController(const pidProfile_t *pidProfile, pidState_t *p
 #endif
 }
 
+void updateMagHoldHeading(int16_t heading)
+{
+    magHoldTargetHeading = heading;
+}
+
+int16_t getMagHoldHeading() {
+    return magHoldTargetHeading;
+}
+
+uint8_t getMagHoldState()
+{
+
+    #ifndef MAG
+        return MAG_HOLD_DISABLED;
+    #endif
+
+    if (!sensors(SENSOR_MAG) || !STATE(SMALL_ANGLE)) {
+        return MAG_HOLD_DISABLED;
+    }
+
+#if defined(NAV)
+    int navHeadingState = naivationGetHeadingControlState();
+    // NAV will prevent MAG_MODE from activating, but require heading control
+    if (navHeadingState != NAV_HEADING_CONTROL_NONE) {
+        // Apply maghold only if heading control is in auto mode
+        if (navHeadingState == NAV_HEADING_CONTROL_AUTO) {
+            return MAG_HOLD_ENABLED;
+        }
+    }
+    else
+#endif
+    if (ABS(rcCommand[YAW]) < 15 && FLIGHT_MODE(MAG_MODE)) {
+        return MAG_HOLD_ENABLED;
+    } else {
+        return MAG_HOLD_UPDATE_HEADING;
+    }
+}
+
+int16_t pidMagHold(const pidProfile_t *pidProfile)
+{
+
+    static filterStatePt1_t magHoldErrorFilter;
+
+    int16_t error = DECIDEGREES_TO_DEGREES(attitude.values.yaw) - magHoldTargetHeading;
+
+    //Apply LPF filter to smoothen response
+    error = filterApplyPt1(error, &magHoldErrorFilter, MAG_HOLD_ERROR_LPF_FREQ, dT);
+
+    if (error <= -180) {
+        error += 360;
+    }
+
+    if (error >= +180) {
+        error -= 360;
+    }
+
+    error *= masterConfig.yaw_control_direction;
+
+    /*
+     * Try limiting diff. If big diff appeared that means that probably this is due to WAYPOINT or RTH fligh mode
+     * Too much diff might cause rapid yaw response and general UAV instability
+     */
+    error = constrain(error, (int) (-1 * masterConfig.mag_hold_heading_diff_limit), masterConfig.mag_hold_heading_diff_limit);
```

Comment:
I would place LPF after this constraint. Currently if error changes from 0 to 180 we still have a jerk here without giving LPF a chance to smooth it.